x/txfees/keeper/hooks.go

```diff
@@ -0,0 +1,163 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	epochstypes "github.com/osmosis-labs/osmosis/v15/x/epochs/types"
+	gammtypes "github.com/osmosis-labs/osmosis/v15/x/gamm/types"
+	poolmanagertypes "github.com/osmosis-labs/osmosis/v15/x/poolmanager/types"
+	"github.com/osmosis-labs/osmosis/v15/x/txfees/types"
+	txfeestypes "github.com/osmosis-labs/osmosis/v15/x/txfees/types"
+)
+
+// Hooks is the wrapper struct for the txfees keeper.
+type Hooks struct {
+	k Keeper
+}
+
+var _ epochstypes.EpochHooks = Hooks{}
+var _ gammtypes.GammHooks = Hooks{}
+
+// Return the wrapper struct
+func (k Keeper) Hooks() Hooks {
+	return Hooks{k}
+}
+
+/* -------------------------------------------------------------------------- */
+/*                                 epoch hooks                                */
+/* -------------------------------------------------------------------------- */
+func (k Keeper) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
+	return nil
+}
+
+// at the end of each epoch, swap all non-DYM fees into DYM and burn them
+func (k Keeper) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
+	if epochIdentifier != k.GetParams(ctx).EpochIdentifier {
+		return nil
+	}
+
+	moduleAddr := k.accountKeeper.GetModuleAddress(txfeestypes.ModuleName)
+	baseDenom, _ := k.GetBaseDenom(ctx)
+
+	//get all balances of this module
+	balances := k.bankKeeper.GetAllBalances(ctx, moduleAddr)
+
+	//swap all to dym
+	for _, coinBalance := range balances {
+		if coinBalance.Denom == baseDenom {
+			continue
+		}
+		if coinBalance.Amount.IsZero() {
+			continue
+		}
+
+		feetoken, err := k.GetFeeToken(ctx, coinBalance.Denom)
+		if err != nil {
+			k.Logger(ctx).Error("unknown fee token", "denom", coinBalance.Denom, "error", err)
+			continue
+		}
+
+		// Do the swap of this fee token denom to base denom.
+		route := []poolmanagertypes.SwapAmountInRoute{
+			{
+				PoolId:        feetoken.PoolID,
+				TokenOutDenom: baseDenom,
+			},
+		}
+		_, err = k.poolManager.RouteExactAmountIn(ctx, moduleAddr, route, coinBalance, sdk.ZeroInt())
+		if err != nil {
+			return err
```

Comment:
if there is an error to swap one of the tokens it will exit the loop and the rest couldn't be switched. Also  the token will still stay in the txFees module to cause problem for every consecutive swap.